contracts/impl/UpgradeImpl.sol

```diff
@@ -0,0 +1,64 @@
+pragma solidity ^0.5.0;
+pragma experimental ABIEncoderV2;
+
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+import "openzeppelin-solidity/contracts/cryptography/ECDSA.sol";
+
+import { EnigmaCommon } from "./EnigmaCommon.sol";
+import { EnigmaState } from "./EnigmaState.sol";
+import { Bytes } from "../utils/Bytes.sol";
+
+/**
+ * @author Enigma
+ *
+ * Library that maintains functionality associated with tasks
+ */
+library UpgradeImpl {
+    using SafeMath for uint256;
+    using SafeMath for uint64;
+    using ECDSA for bytes32;
+    using Bytes for address;
+
+    function upgradeEnigmaContractImpl(
+        EnigmaState.State storage state,
+        address _updatedEnigmaContractAddress
+    )
+    public
+    {
+        state.updatedEnigmaContractAddress = _updatedEnigmaContractAddress;
+
+        for (uint i = 0; i < state.taskIds.length; i++) {
+            if (state.tasks[state.taskIds[i]].status == EnigmaCommon.TaskStatus.RecordCreated) {
+                EnigmaCommon.TaskRecord storage task = state.tasks[state.taskIds[i]];
+
+                // Return the full fee to the task sender
+                require(state.engToken.transfer(task.sender, task.gasLimit.mul(task.gasPx)), "Token transfer failed");
```

Comment:
So this means that if a tokenTransfer fails, the upgrade does not happen. Can we document the possible causes of this, and confirm that this is the behavior we want?
 
Is the only reason that this may fail that it runs out of gas?

At least, we agree that if task `n` fails, and the upgrade aborts, we would be able to trigger an upgrade again, and the first `n-1` tasks were already processed, and the tokenTransfer process would start again from where it left off at `n`.

My question is if there is any scenario in which the upgrade gets stuck at taksk `n` for any reason (Cc: @fredfortier)

Reply:
I can't foresee why it might get stuck. My overall concern with the way that we handle gas is that the ENG gas might be small than the ETH gas required for the `transfer` call. That's why I suggested that we have user balances and a withdrawal fn. That said, this condition should not cause this batch fn to fail as long as enough gas is provided.